source/common/network/connection_impl.cc

```diff
@@ -432,12 +432,11 @@ void ConnectionImpl::onReadReady() {
   // If this connection doesn't have half-close semantics, translate end_stream into
   // a connection close.
   if ((!enable_half_close_ && result.end_stream_read_)) {
-    result.end_stream_read_ = false;
     result.action_ = PostIoAction::Close;
   }
 
   read_end_stream_ |= result.end_stream_read_;
-  if (result.bytes_processed_ != 0 || result.end_stream_read_) {
+  if (result.bytes_processed_ != 0 || (enable_half_close_ && result.end_stream_read_)) {
```

Comment:
If end_stream is delivered without any data, and you want this to be delivered, I think you want to revert this line. Otherwise end_stream==true will only be set if it arrives with some data (or enable_half_close_ is true).

Reply:
> end_stream==true will only be set if it arrives with some data (or enable_half_close_ is true).

This is intentional, since we have converted half-close to full-close above, raising onData(0-length buffer, true) for non-half-close read filter is not needed since it will be destroyed right after `onReadReady`.